spec/index.html

```diff
@@ -1255,7 +1311,25 @@ <h3>Algorithm</h3>
           </details>
         </li>
         <li id="ca.7">Return the <a>serialized canonical form</a>
-          of the <a>normalized dataset</a>.</li>
+          of the <a>normalized dataset</a>.
+          Alternatively, return the <a>normalized dataset</a> itself,
+          which includes the <a>input blank node identifier map</a>,
+          and <a>canonical issuer</a>.
```

Comment:
```suggestion
          Optionally, the algorithm may also return the <a>normalized dataset</a> as an auxiliary output,
          which includes the <a>input blank node identifier map</a>,
          and <a>canonical issuer</a>.
```
How about saying that the *output* of the c14n (single and deterministic) is the serialized form, whereas the normalized dataset (possibly non-deterministic) can be obtained as an *auxiliary output*?

Reply:
So, would change "Optionally" to "As an auxiliary output" satisfy this? I don't really follow how this is non-deterministic, as it would seem that with a given input, the same normalized dataset would be produced.

Reply:
I'd rather not be so prescriptive in saying that you have to return both -- I'd be happy with an either / or. I don't think we should say what implementations can do as a proxy for indicating that two different implementations might technically map one blank node to ID A and another implementation might map it to ID B. This is @yamdan's point, I believe -- but this only happens when there are isomorphisms that make this difference irrelevant. A serialized version of the dataset would look the same.

Perhaps that's what we say in a note: "Technically speaking, one implementation might map particular blank nodes to different identifiers than another implementation, however, this only occurs when there are isomorphisms in the dataset such that a serialized expression of the dataset would appear the same from either implementation."

And then we can say that algorithms may return both the canonically serialized dataset and the normalized dataset or either of these as requested by the invoker of the algorithm.

Reply:
It might be some time before I can do this update, but it seems simple enough. I'm traveling for the next week, and internet access is spotty. Feel free to update and commit, as this is really just informative, now.

Reply:
I've added a suggestion below.

Reply:
@gkellogg,
> I don't really follow how this is non-deterministic, as it would seem that with a given input, the same normalized dataset would be produced.

**Non-deterministic** choice may occur in step 5.3 of the [4.4.3 Algorithm](https://w3c.github.io/rdf-canon/spec/#canon-algo-algo), where it's possible to have ties of *result* in the *hash path list* so that it's non-deterministic which *result* is firstly chosen from the list.
Even the same implementation can output different canonical issuers depending on the runtime environment or the input blank node identifiers.

The only thing I would like to eliminate here is the possibility of creating a misuse of the *normalized dataset*, believing that it is a deterministic and single canonical result and connecting it to the hash or signature input.
I think we can prevent this by clearly stating that the serialized form is the output of the canonicalization and the normalized dataset is an auxiliary output.

As @dlongley mentioned, I think this only happens when there are automorphisms in the input dataset.

Reply:
I will continue this "non-deterministic" topic in a new separate PR.